### DISCUSSION TITLE
RNNModel is not working

### DISCUSSION THREAD

**Discussion by Pedklin** [etna-team/etna#457] [Q&A] [answered]
Никак не могу добиться, чтобы модель дала результат...
Код скопировал из вашей документации: https://docs.etna.ai/stable/tutorials/202-NN_examples.html.
Небольшие отличия из-за особенностей датасета (файл с датасетом приложил):
original_df = pd.read_csv("salary2.csv")
[salary2.csv](https://github.com/user-attachments/files/16759924/salary2.csv)

original_df["timestamp"] = pd.to_datetime(original_df["week_key"])
original_df["target"] = original_df["sal_median"]
original_df["segment"] = original_df["segm"]
original_df.drop(columns=["week_key", "sal_median", "sal_diff", "segm"], inplace=True)
df = TSDataset.to_dataset(original_df)
ts = TSDataset(df, freq="W-SUN")            // freq отличается от примера в документации, там D
horizon = 52
num_lags = 5

Насколько я понял embedding_sizes, поскольку в данном примере 2 сегмента, заменил определение так:
embedding_sizes = {"dateflag_day_number_in_week": (7, 7), "segment_code": (2, 7)}

    model = RNNModel(
    input_size = num_lags + 1,
    encoder_length = 2 * horizon,
    decoder_length = horizon,
    embedding_sizes = embedding_sizes,
    trainer_params = dict(max_epochs = 5), 
    lr=1e-3)
    transforms=[scaler, transform_lag, transform_date, segment_encoder, label_encoder]
    pipe = Pipeline(model=model, transforms=transforms, horizon=horizon)
    metric_tools = []
    metric_tools.append(SMAPE())
    metric_tools.append(MSE())
    metrics, forecast, info = pipe.backtest(ts, metrics=metric_tools, n_folds=1)
    
    При таких условиях pipe.backtest падает. Если уменьшить horizon до 45, не падает, но forecast даже близко не похож.
    Я пробовал, поскольку здесь freq="W-SUN", заменить на
    transform_date = DateFlagsTransform(day_number_in_week=False, day_number_in_month=True, out_column="dateflag")
    label_encoder = LabelEncoderTransform(
    in_column="dateflag_day_number_in_month", strategy="none", out_column="dateflag_day_number_in_month_label")
    embedding_sizes = {"dateflag_day_number_in_month": (4, 4), "segment_code": (2, 4)}
    В этом случае падает.

**Comment by egoriyaa** [comment]
Hi!
Can you share your full code snippet and error output?

**Reply by Pedklin** [reply]
Hi!
Full code:
original_df = pd.read_csv("salary2.csv")
original_df["timestamp"] = pd.to_datetime(original_df["week_key"])
original_df["target"] = original_df["sal_median"]
original_df["segment"] = original_df["segm"]
original_df.drop(columns=["week_key", "sal_median", "sal_diff", "segm"], inplace=True)
df = TSDataset.to_dataset(original_df)
ts = TSDataset(df, freq="W-SUN") // freq отличается от примера в документации, там D
horizon = 52
num_lags = 10
scaler = StandardScalerTransform(in_column="target")
transform_lag = LagTransform(
in_column="target",
lags=[horizon + i for i in range(num_lags)],
out_column="target_lag" )
transform_date = DateFlagsTransform(day_number_in_week=True, day_number_in_month=False, out_column="dateflag")
segment_encoder = SegmentEncoderTransform()
label_encoder = LabelEncoderTransform(
in_column="dateflag_day_number_in_week", strategy="none", out_column="dateflag_day_number_in_week_label")
embedding_sizes = {"dateflag_day_number_in_week": (7, 7), "segment_code": (2, 7)}

seed = 42
random.seed(seed)
numpy.random.seed(seed)
torch.manual_seed(seed)
torch.cuda.manual_seed_all(seed)

model = RNNModel(
input_size = num_lags + 1,
encoder_length = 2 * horizon,
decoder_length = horizon,
embedding_sizes = embedding_sizes,
trainer_params = dict(max_epochs = 10), 
lr=1e-3)

transforms=[scaler, transform_lag, transform_date, segment_encoder, label_encoder]

pipe = Pipeline(model=model, transforms=transforms, horizon=hrz)
metric_tools = []
metric_tools.append(SMAPE())
metric_tools.append(MSE())
metrics, forecast, info = pipe.backtest(ts, metrics=metric_tools, n_folds=1)
plot_backtest(forecast_df=forecast, ts=ts, history_len=104)
 
pipe.backtest falls with error:
num_samples should be a positive integer value, but got num_samples = 0

if we change horizon = 45 then there are no errors but we have the following owfull result:
![rnn](https://github.com/user-attachments/assets/3be02ffa-6403-4779-97a2-0026be139746)

**Reply by egoriyaa** [reply]
1. Error raises because number of timestamps is less then needed for your RNN config (encoder_length + decoder_length)
2. As to bad results, most probably you have not enough data (returning to the first answer) or not enough epochs  to train good model

**Reply by Pedklin** [reply]
1. In this example number of timestamps = 228 > encoder_length + decoder_length = 156...
2. Well, ok, i reduce horizon = 20.
I tryed to make following enhancement: if our frequency is W-SUN so i changed    

transform_date = DateFlagsTransform(day_number_in_week=False, day_number_in_month=True, out_column="dateflag")
label_encoder = LabelEncoderTransform(in_column="dateflag_day_number_in_month", strategy="none", 
                                                                 out_column="dateflag_day_number_in_month_label")
embedding_sizes = {"dateflag_day_number_in_month": (31, 31), "segment_code": (2, 31)}

And also i set num_lags = 52. Should num_lags correlates with the seasonality?
So i got the following result:
 ![rnn](https://github.com/user-attachments/assets/9287a2c6-4491-436c-a430-c9e72cdb25e5)
May be it is not too bad....

**Reply by egoriyaa** [reply]
Okey, to be more precise, your data is splitted in the following way:
- last `horizon` steps are test part, because you set `n_folds=1`
- first `horizon + num_lags` steps are deleted because `lag_{horizon+num_lags}` feature is NaN and model can't be trained
- so there are `228 - horizon - (horizon + num_lags) + 1 = 115` remaining steps, but min number of timestamps is `encoder_length + decoder_length = 156`

**Reply by egoriyaa** [reply, marked as answer]
As to number of lags:
Ideally, the number of lags is selected using a backtest, but if there are few rows, it is easy to overfit. As an initial approximation, I would consider the reference points: the length of the seasonality, the horizon
The distance between two "peaks" looks like seasonality.

**Comment by Pedklin** [comment]
Next problem with this model:
at the row
metrics, forecast, info = pipe.backtest(ts, metrics=metric_tools, n_folds=1)
I get the error
input.size(-1) must be equal to input_size. Expected 35, got 36
This error rises only if i run model in environment created by PyInstaller. When the same script runs in VSCode it is ok.
PyInstaller:
[ETHNA.zip](https://github.com/user-attachments/files/17231073/ETHNA.zip)

**Reply by egoriyaa** [reply]
Hello! Provide a full error trace, please.

**Reply by Pedklin** [reply]
Data file:
[114-124.csv](https://github.com/user-attachments/files/17231375/114-124.csv)
Frankly i do not know how to get trace from exe....

**Reply by egoriyaa** [reply]
1. We found a small bug in example notebook with NN models (see #488) (maybe it will help)
2. It is hard to understand the core problem without understanding what library/function raises error. Try to found previous output before `input.size(-1) must be equal to input_size. Expected 35, got 36`

**Reply by egoriyaa** [reply]
I found out that this error raises when the total number of features is not equal to expected number by model layers.
In your case there is one more feature then expected. In a quick glance I can't understand how many features there should be because there are many ifs.

**Reply by Pedklin** [reply]
Yes, thank you, the problem was that ts = TSDataset(df, freq=freq) contained extra columns that i did not drop.
When i dropped them then now all is ok.
